pacman/operations/routing_info_allocator_algorithms/zoned_routing_info_allocator.py

```diff
@@ -16,154 +16,308 @@
+            by_app = by_app_and_partition_name[app_id]
+            for partition_name, by_partition_name in by_app.items():
+                max_keys = 0
+                for mac_vertex in by_partition_name:
+                    partition = self.__machine_graph.\
+                        get_outgoing_edge_partition_starting_at_vertex(
+                            mac_vertex, partition_name)
+                    n_keys = self.__n_keys_map.n_keys_for_partition(partition)
+                    max_keys = max(max_keys, n_keys)
+                if max_keys > 0:
+                    atom_bits = self.__bits_needed(max_keys)
+                    self.__n_bits_atoms = max(self.__n_bits_atoms, atom_bits)
+                    machine_bits = self.__bits_needed(len(by_partition_name))
+                    self.__n_bits_machine = max(
+                        self.__n_bits_machine, machine_bits)
+                    self.__n_bits_atoms_and_mac = max(
+                        self.__n_bits_atoms_and_mac, machine_bits + atom_bits)
+                    self.__atom_bits_per_app_part[
+                        (app_id, partition_name)] = atom_bits
+
+    def __check_zones(self):
+        # See if it could fit even before considerding fixed
+        app_part_bits = self.__bits_needed(len(self.__atom_bits_per_app_part))
+        if app_part_bits + self.__n_bits_atoms_and_mac > BITS_IN_KEY:
             raise PacmanRouteInfoAllocationException(
                 "Unable to use ZonedRoutingInfoAllocator please select a "
                 "different allocator as it needs {} + {} bits".format(
-                    source_bits, max_app_keys_bits))
-        if max_partitions > self.MAX_PARTITIONS_SUPPORTED:
-            raise NotImplementedError()
-        return max_app_keys_bits, key_bits_per_app
+                    app_part_bits, self.__n_bits_atoms_and_mac))
 
-    def _simple_allocate(self, max_app_keys_bits, key_bits_map):
+        # Reserve fixed and check it still works
+        self.__check_fixed()
+        app_part_bits = self.__bits_needed(
+            len(self.__atom_bits_per_app_part) + len(self.__fixed_used))
+        if app_part_bits + self.__n_bits_atoms_and_mac > BITS_IN_KEY:
+            raise PacmanRouteInfoAllocationException(
+                "Unable to use ZonedRoutingInfoAllocator please select a "
+                "different allocator as it needs {} + {} bits".format(
+                    app_part_bits, self.__n_bits_atoms_and_mac))
+
+        if not self.__flexible:
+            if app_part_bits + self.__n_bits_machine + self.__n_bits_atoms > \
+                    BITS_IN_KEY:
+                # We know from above test that all will fit if flexible
+                # Reduce the suggested size of n_bits_atoms
+                self.__n_bits_atoms = \
+                    BITS_IN_KEY - app_part_bits - self.__n_bits_machine
+                self.__n_bits_atoms_and_mac = BITS_IN_KEY - app_part_bits
```

Comment:
I think this just means that it doesn't fit... how can you reduce the number of bits used by the atoms if we need to keep the mask the same?

Reply:
Looking again at this, I can't help but think it will go wrong then you get one of the vertices whose atoms happen to be bigger than can be represented by n_bits_atoms...

Reply:
OK, I can see how this could help; if you can't fit the global, you make it slightly more flexible.  It might help, but it might not...  I guess a warning could be useful in this case?